stream-chat-android-state/src/test/java/io/getstream/chat/android/offline/internal/TypingEventPrunerTest.kt

```diff
@@ -0,0 +1,168 @@
+        typingEventPruner.initialize(channelId)
+
+        val typingStartEvent = randomTypingStartEvent()
+
+        // registerListener
+        typingEventPruner.setOnTypingEventsUpdatedListener(onTypingUpdateListenerMock)
+
+        // Should add the typing event
+        typingEventPruner.processEvent(typingStartEvent.user.id, typingStartEvent)
+        // Should remove the previously added typing event
+        typingEventPruner.processEvent(typingStartEvent.user.id, null)
+
+        // Called once after event addition, once after removal
+        verify(onTypingUpdateListenerMock, times(2))
+    }
+
+    @Test
+    fun `Given multiple typing event from the same user Should output a collection with a single latest typing event`() =
+        runTest {
+            val typingEventPruner = TypingEventPruner(this)
+            typingEventPruner.initialize(channelId)
+
+            val firstTypingStartEvent = randomTypingStartEvent()
+
+            // Create a copy of the first event by simulating an event from the same user at a later date
+            val advancedStartEventCreationDate = Date(firstTypingStartEvent.createdAt.time + 100)
+            val secondTypingStartEvent = firstTypingStartEvent.copy(createdAt = advancedStartEventCreationDate)
+
+            // Set both events
+            typingEventPruner.processEvent(firstTypingStartEvent.user.id, firstTypingStartEvent)
+            typingEventPruner.processEvent(secondTypingStartEvent.user.id, secondTypingStartEvent)
+
+            typingEventPruner.getRawTyping() shouldBeEqualTo mapOf(
+                Pair(
+                    secondTypingStartEvent.user.id,
+                    secondTypingStartEvent
+                )
+            )
+            typingEventPruner.getTypingEvent() shouldBeEqualTo TypingEvent(
+                channelId,
+                listOf(secondTypingStartEvent.user)
+            )
+        }
+
+    @Test
+    fun `Given typing events that are not manually cancelled Should be automatically cleaned by TypingEventPruner`() =
+        runTest {
+            val prunerDelay = TypingEventPruner.DEFAULT_DELAY_TIME_MS
+            val typingEventPruner = TypingEventPruner(this, prunerDelay)
+            typingEventPruner.initialize(channelId)
+
+            val firstTypingStartEvent = randomTypingStartEvent(user = randomUser(id = "User1"))
+            val secondTypingStartEvent = randomTypingStartEvent(user = randomUser(id = "User2"))
+
+            typingEventPruner.processEvent(firstTypingStartEvent.user.id, firstTypingStartEvent)
+            typingEventPruner.processEvent(secondTypingStartEvent.user.id, secondTypingStartEvent)
+
+            // Time is advanced by slightly more than it takes for the pruner to remove stale events
+            advanceTimeBy(prunerDelay + 1)
+
+            typingEventPruner.getRawTyping() shouldBeEqualTo mapOf()
+            typingEventPruner.getTypingEvent() shouldBeEqualTo TypingEvent(channelId, listOf())
```

Comment:
Can't we "verify" over the `onUpdated` listener?
It should be the same, and we hide the internal implementation

Reply:
Yes, that's a good idea, we can do that. :smile:

Reply:
Done :smile: